x/market/internal/keeper/swap.go

```diff
@@ -0,0 +1,151 @@
+// Returns an Error if the swap is recursive, or the coins to be traded are unknown by the oracle, or the amount
+// to trade is too small.
+func (k Keeper) ComputeSwap(ctx sdk.Context, offerCoin sdk.Coin, askDenom string) (retDecCoin sdk.DecCoin, spread sdk.Dec, err sdk.Error) {
+
+	// Return invalid recursive swap err
+	if offerCoin.Denom == askDenom {
+		return sdk.DecCoin{}, sdk.ZeroDec(), types.ErrRecursiveSwap(k.codespace, askDenom)
+	}
+
+	// Swap offer coin to base denom for simplicity of swap process
+	baseOfferDecCoin, err := k.ComputeInternalSwap(ctx, sdk.NewDecCoinFromCoin(offerCoin), core.MicroSDRDenom)
+	if err != nil {
+		return sdk.DecCoin{}, sdk.Dec{}, err
+	}
+
+	// Get swap amount based on the oracle price
+	retDecCoin, err = k.ComputeInternalSwap(ctx, baseOfferDecCoin, askDenom)
+	if err != nil {
+		return sdk.DecCoin{}, sdk.Dec{}, err
+	}
+
+	// TERRA<->TERRA swap
+	// Apply only tobin tax without constant product spread
+	if offerCoin.Denom != core.MicroLunaDenom && askDenom != core.MicroLunaDenom {
+		spread = k.TobinTax(ctx)
+		return
+	}
+
+	eqTerraPool := k.GetEqTerraPool(ctx)
+	minSpread := k.MinSpread(ctx)
+
+	var offerPoolInfo types.PoolInfo // delta is base denom(usdr) unit
+	var askPoolInfo types.PoolInfo   // delta is base denom(usdr) unit
+	var offerPool sdk.Dec            // base denom(usdr) unit
+	var askPool sdk.Dec              // base denom(usdr) unit
+	if offerCoin.Denom != core.MicroLunaDenom {
+		// TERRA<->LUNA swap
+		offerPoolInfo = k.GetPoolInfo(ctx, core.MicroSDRDenom)
+		offerPool = eqTerraPool.Add(offerPoolInfo.Delta)
+		askPoolInfo = k.GetPoolInfo(ctx, core.MicroLunaDenom)
+		askPool = eqTerraPool.Add(askPoolInfo.Delta)
+	} else {
+		// LUNA<->TERRA swap
+		offerPoolInfo = k.GetPoolInfo(ctx, core.MicroLunaDenom)
+		offerPool = eqTerraPool.Add(offerPoolInfo.Delta)
+		askPoolInfo = k.GetPoolInfo(ctx, core.MicroSDRDenom)
+		askPool = eqTerraPool.Add(askPoolInfo.Delta)
+	}
+
+	// constant-product, which by construction is square of equilibrium Terra pool
+	cp := eqTerraPool.Mul(eqTerraPool)
+
+	// Get cp based swap amount
+	// askBaseAmount = askPool - cp / (offerPool + offerBaseAmount)
+	// askBaseAmount is base denom(usdr) unit
+	askBaseAmount := askPool.Sub(cp.Quo(baseOfferDecCoin.Amount.Add(offerPool)))
+
+	// Swap base coin to ask denom
+	askDecCoin, err := k.ComputeInternalSwap(ctx, sdk.NewDecCoinFromDec(core.MicroSDRDenom, askBaseAmount), askDenom)
+	if err != nil {
+		return sdk.DecCoin{}, sdk.ZeroDec(), err
+	}
+
+	// spread = max(contant_product_spread + tobin_tax, tobin_tax)
+	// contant_product_spread can be negative
+	askDecAmount := askDecCoin.Amount
+	retDecAmount := retDecCoin.Amount
+	spread = retDecAmount.Sub(askDecAmount).Quo(retDecAmount).Add(minSpread)
+
+	if spread.LT(minSpread) {
+		spread = minSpread
+	}
+
+	return
+}
+
+// ComputeInternalSwap returns the amount of asked DecCoin should be returned for a given offerCoin at the effective
```

Comment:
Do we still need this function?

Reply:
Not necessary, but for convenience. 
In swap process and treasury module, ComputeInternalSwap is used to convert a denom to another at the effective exchange rate without spread.